glib/src/object.rs

```diff
@@ -1525,9 +1525,78 @@ impl<'a, O: IsA<Object> + IsClass> ObjectBuilder<'a, O> {
     }
 
     // rustdoc-stripper-ignore-next
-    /// Set property `name` to the given value `value`.
+    /// Build the object with the provided properties.
+    ///
+    /// # Panics
+    ///
+    /// This panics if the object is not instantiable, doesn't have all the given properties or
+    /// property values of the wrong type are provided.
+    #[track_caller]
+    #[inline]
+    pub fn build(mut self) -> O {
+        let object = Object::with_mut_values(self.type_, &mut self.properties);
+        unsafe { object.unsafe_cast::<O>() }
+    }
+}
+
+pub trait ObjectBuilderPropertySetter<'a>: Sized {
+    // rustdoc-stripper-ignore-next
+    /// Sets property `name` to the given value `value`.
+    ///
+    /// Overrides any default or previously defined value.
+    fn property(self, name: &'a str, value: impl Into<Value>) -> Self;
+
+    // rustdoc-stripper-ignore-next
+    /// Sets property `name` to the given inner value if `value` is `Some`.
+    ///
+    /// This has no effect if the value is `None`, i.e. default or previous value is kept.
+    #[inline]
+    fn property_if_some(self, name: &'a str, value: Option<impl Into<Value>>) -> Self {
+        if let Some(value) = value {
+            ObjectBuilderPropertySetter::property(self, name, value)
+        } else {
+            self
+        }
+    }
+
+    // rustdoc-stripper-ignore-next
+    /// Sets property `name` using the given `ValueType` `V` built from `iter`'s the `Item`s.
+    ///
+    /// Overrides any default or previously defined value.
     #[inline]
-    pub fn property(self, name: &'a str, value: impl Into<Value>) -> Self {
+    fn property_from_iter<V: ValueType + Into<Value> + FromIterator<SendValue>>(
+        self,
+        name: &'a str,
+        iter: impl IntoIterator<Item = impl Into<SendValue>>,
+    ) -> Self {
+        let iter = iter.into_iter().map(|item| item.into());
+        ObjectBuilderPropertySetter::property(self, name, V::from_iter(iter))
+    }
+
+    // rustdoc-stripper-ignore-next
+    /// Sets property `name` using the given `ValueType` `V` built from `iter`'s Item`s,
+    /// if `iter` is not empty.
+    ///
+    /// This has no effect if `iter` is empty, i.e. previous property value is unchanged.
+    #[inline]
+    fn property_if_any<V: ValueType + Into<Value> + FromIterator<SendValue>>(
+        self,
+        name: &'a str,
+        iter: impl IntoIterator<Item = impl Into<SendValue>>,
+    ) -> Self {
+        let mut iter = iter.into_iter().peekable();
+        if iter.peek().is_some() {
+            let iter = iter.map(|item| item.into());
+            ObjectBuilderPropertySetter::property(self, name, V::from_iter(iter))
+        } else {
+            self
+        }
+    }
+}
+
+impl<'a, O: IsA<Object> + IsClass> ObjectBuilderPropertySetter<'a> for ObjectBuilder<'a, O> {
```

Comment:
Why is this an extension trait instead of just implementing it directly on `ObjectBuilder`?

Reply:
Because you suggested this :): https://gitlab.freedesktop.org/gstreamer/gstreamer-rs/-/merge_requests/1431#note_2392609

I think it's convenient to have a single definition too: I can use this `trait` for `gst::ElementFactory`, implement `property()` and get all the other implementations for free.

I used a similar pattern `gst::StructureBuilderPropertySetter` which is applicable to `gst::StructureBuilder`, `gst::CapsBuilder`, `gst_audio::CapsBuilder`, gst_video::CapsBuilder`, ...

Reply:
I suggested this because I thought of adding this extension trait in gstreamer-rs, but here it could also be a normal impl block.

If it makes anything easier and you're going to implement the trait on other types too that's fine. (Btw, could you also add this to the appsrc/appsink builders then? I just noticed that they don't have the general `property` method)

Reply:
> I can only see one implementation here

But more in gstreamer-rs?

And here, I think all the autogenerated builders for specific types could make use of this. I think they all don't have a generic property method right now but that would be useful.

Reply:
I'm considering using a macro instead of a `trait`. It would be possible to add `xxx_if()`,  `xxx_if_some()`,  `xxx_from_iter()` &  `xxx_if_not_empty()` to any builder implementing `xxx()`. There would be a specific macro in GStreamer due to the `Send` constraint, but it would be applicable for `ElementFactory` (property), `Structure` (field),  `Caps` (field), `Event` (other_field) & `Message` (other_field).

Reply:
A macro that is exported from the glib crate? Not sure I like adding more macro public API. @bilelmoussaoui ?

Reply:
The way I see it: due to the `Value` / `SendValue` differences, I will handle this separately in `gtk-rs` & `GStreamer`.

* In `gtk-rs`: the new methods `property_*` can be added to the `ObjectBuilder` `impl` directly. For the generated `Builder`s, a `gir` update could add the additional methods on a per-property basis.
* In `GStreamer`, I think I can use the macro I was mentioning in my previous comment. It would be transparent to users since it would only be used in the bindings for the `ElementFactory` & the `Structure`-like `struct` builders, not something like `glib::clone!` for instance.

Reply:
Sounds like a plan

Reply:
I don't see which other types that could use this. I think the only use case would be either gstreamer-rs or gegl-rs(which is not maintained anyways...). So do we really need this to be an extension trait?

Reply:
We don't. See my comment: https://github.com/gtk-rs/gtk-rs-core/pull/1377#discussion_r1586068809

Reply:
Related gir issue: https://github.com/gtk-rs/gir/issues/1581